bfxapi/examples/ws/multiple_instances.py

```diff
@@ -0,0 +1,181 @@
+        subbed_tickers = get_all_subscriptions_tickers()
+        print(f'Subscribed tickers: {subbed_tickers}')
+
+        # if ticker is not in subbed tickers, then we subscribe to the channel
+        to_sub = [f"t{ticker}" for ticker in get_random_list_of_tickers() if ticker not in subbed_tickers]
+        for ticker in to_sub:
+            print(f'To subscribe: {ticker}')
+            instance = get_available_instance()
+            if instance and instance.is_ready:
+                print(f'Subscribing on instance {instance.id}')
+                await instance.subscribe([ticker])
+            else:
+                instances_to_create = math.ceil(len(to_sub) / MAX_CHANNELS)
+                create_instances(instances_to_create)
+                break
+
+        to_unsub = [f"t{ticker}" for ticker in subbed_tickers if ticker in get_random_list_of_tickers()]
+        if len(to_unsub) > 0:
+            print(f'To unsubscribe: {to_unsub}')
+            for instance in instances:
+                await instance.unsubscribe(to_unsub)
+
+    def stop(self):
+        self.task.cancel()
+        self.is_stopped = True
+
+
+instances = []
+
+
+def get_all_subscriptions_tickers():
+    tickers = []
+    for instance in instances:
+        for ticker in instance.subscriptions['trades']:
+            tickers.append(ticker)
+    return tickers
+
+
+def count_open_channels(instance):
+    return len(instance.subscriptions['trades']) + len(instance.subscriptions['ticker'])
+
+
+def create_instances(instances_to_create):
+    for _ in range(0, instances_to_create):
+        instance = Instance(len(instances))
+        instance.run()
+        instances.append(instance)
+
+
+def get_available_instance():
+    for instance in instances:
+        if count_open_channels(instance) + 1 <= MAX_CHANNELS:
+            return instance
+    return None
+
+
+def log_error(err):
+    print("Error: {}".format(err))
+
+
+def log_trade(trade):
+    print(trade)
+
+
+def log_ticker(ticker):
+    print(ticker)
+
+
+async def on_subscribe(instance, subscription):
+    print(f'Subscribed to {subscription.symbol} channel {subscription.channel_name}')
+    instance.subscriptions[subscription.channel_name][subscription.symbol] = subscription.chan_id
+
+
+async def on_unsubscribed(instance, subscription):
+    print(f'Unsubscribed to {subscription.symbol} channel {subscription.channel_name}')
+    instance.subscriptions[subscription.channel_name][subscription.symbol] = subscription.chan_id
+    del instance.subscriptions[subscription.channel_name][subscription.symbol]
+
+
+async def on_connected(instance):
+    print(f"Instance {instance.id} is connected")
+    instance.is_ready = True
+
+
+async def on_stopped(instance):
+    print(f"Instance {instance.id} is dead, removing it from instances list")
+    instances.pop(instance.id)
+
+
+def run():
+    loop = asyncio.get_event_loop()
+    task = Routine(loop, ws, interval=5)
+    loop.run_until_complete(task)
+
+run()
```

Comment:
missing newline on end of file

(you can set your editor to automatically add them)

Reply:
ty

Reply:
done